src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,253 @@
+  if (U_SUCCESS(status)) {
+    return out.slice(0, target - out.asChars().begin()).attach(kj::mv(out));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeToUCS2(kj::ArrayPtr<kj::byte> source, Encoding fromEncoding,
+                                               Encoding toEncoding) {
+  UErrorCode status{};
+  const size_t length_in_chars = source.size() * sizeof(UChar);
+  Converter from(fromEncoding);
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(source.size() <= 134217728, Error, "Source buffer is too large to transcode");
+  auto destbuf = kj::heapArray<UChar>(source.size());
+  ucnv_toUChars(from.conv(), destbuf.begin(), length_in_chars,
+                const_cast<const char*>(source.asChars().begin()), source.size(), &status);
+
+  if (U_SUCCESS(status)) {
+    return destbuf.asBytes().attach(kj::mv(destbuf));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeFromUCS2(kj::ArrayPtr<kj::byte> source,
+                                                 Encoding fromEncoding, Encoding toEncoding) {
+  UErrorCode status{};
+  Converter to(toEncoding);
+  auto substitude = kj::str(kj::repeat('?', to.minCharSize()));
+  to.setSubstitudeChars(substitude);
+
+  const size_t length_in_chars = source.size() * sizeof(UChar);
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(length_in_chars <= 134217728, Error, "Buffer is too large to transcode");
+
+  // TODO: Optimization opportunity: Remove sourcebuf allocation by calculating length beforehand.
+  KJ_STACK_ARRAY(UChar, sourcebuf, source.size(), 1024, 536870888);
+  memcpy(sourcebuf.begin(), sourcebuf.begin(), length_in_chars);
+
+  auto destbuf = kj::heapArray<kj::byte>(length_in_chars);
+  auto len = ucnv_fromUChars(to.conv(), destbuf.asChars().begin(), length_in_chars,
+                             const_cast<const UChar*>(sourcebuf.begin()), length_in_chars, &status);
+
+  if (U_SUCCESS(status)) {
+    return destbuf.slice(0, len).attach(kj::mv(destbuf));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeUcs2FromUtf8(kj::ArrayPtr<kj::byte> source,
+                                                     Encoding fromEncoding, Encoding toEncoding) {
+  size_t expected_utf16_length =
+      simdutf::utf16_length_from_utf8(source.asChars().begin(), source.size());
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(expected_utf16_length <= 134217728, Error,
+              "Expected UTF-16le length is too large to transcode");
+  auto destbuf = kj::heapArray<UChar>(expected_utf16_length);
+
+  size_t actual_length =
+      simdutf::convert_utf8_to_utf16le(source.asChars().begin(), source.size(), destbuf.begin());
+  KJ_ASSERT(actual_length == expected_utf16_length, "Expected UTF16 length mismatch");
```

Comment:
To be safest, convert these `KJ_ASSERT` uses into `JSG_REQUIRE` so that we get an actual JS error thrown and not the dreaded "Internal error"